File: samples/fl/implementations/java/PrettyPrinter.java
import types.*;

/**
 * A visitor for pretty printing
 */
public class PrettyPrinter extends Visitor {
	//
	// An AGGREGATING string variable is potentially very inefficient.
	//
	private String result = "";
	public String getResult() {
		return result;
	}
	public void visit(Program x) {
		for (Function f : x.functions) {
			f.accept(this);
			result += "\n";
		}
	}
	public void visit(Function x) {
		result += x.name;
		for (String s : x.args) {
			result += " " + s;
		}
		result += " = ";
		x.rhs.accept(this);
	}
	public void visit(Literal x) {
		result += x.info;
	}
	public void visit(Argument x) {
		result += x.name;
	}
	public void visit(Binary x) {
		result += "(";
		x.left.accept(this);
		switch (x.ops) {
		    case Equal :
			result += "==";
			break;
		    case Plus :
			result += " + ";
			break;
		    case Minus :
			result += " - ";
			break;
		}
		x.right.accept(this);		
		result += ")";
	}
	public void visit(Apply x) {
		result += "(" + x.name;
		for (Expr e : x.args) {
			result += " ";
			e.accept(this);
		}
		result += ")";
	}
	public void visit(IfThenElse x) {
		result += "if ";
		x.exprIf.accept(this);
		result += " then ";
		x.exprThen.accept(this);
		result += " else ";
		x.exprElse.accept(this);
	}
}

File: topics/fl/implementations/haskell/PrettyPrinter.hs
module PrettyPrinter where

import Types

-- Pretty-print to file

ppToFile f x = do
        let s = prettyPrint x
	writeFile f s

class Pp x where prettyPrint :: x -> String

instance Pp x => Pp [x]
  where
    prettyPrint  = concat . map ((++"\n") . prettyPrint)

instance Pp Function
  where
    prettyPrint (Function n ns e) =
    	 concat [n, concat (map (\n -> " " ++ n) ns)," = ",prettyPrint e]

instance Pp Expr
  where
    prettyPrint (Literal i) = show i
    prettyPrint (Argument n) = n
    prettyPrint (Binary o x y) = concat ["(",prettyPrint x,prettyPrint o,prettyPrint y,")"]
    prettyPrint (IfThenElse x y z) = concat ["if ",prettyPrint x," then ",prettyPrint y," else ",prettyPrint z]
    prettyPrint (Apply n es) = concat ["(",concat (n:map (\e -> " " ++ prettyPrint e) es),")"]

instance Pp Ops
  where
    prettyPrint Equal = "=="
    prettyPrint Plus = " + "
    prettyPrint Minus = " - "
